Bug 1239965 - Door hanger prompts the user to share on Hello as he copies the URL content

Mardak opened the pull request.

Mardak committed:
Bug 1239965 - Door hanger prompts the user to share on Hello as he copies the URL content. r=dmose

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -346,6 +348,132 @@ var WindowListener = {
+        this._dns.asyncResolve(
+          Services.prefs.getCharPref("loop.copy.throttler"),
+          this._dns.RESOLVE_DISABLE_IPV6,
+          (request, record) => {
```
Please break the callback out into its own named function for improved readability of the parent function.

Mardak replied:
Moved to `let onDNS` with comments. Also collapsed the multiline-argument-style asyncResolve call to singleline-argument-style.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -346,6 +348,132 @@ var WindowListener = {
+            null, // Origin
+            "chrome://loop/content/panels/copy.html",
+            null, // Size
+            iframe => {
```
Please also break this callback out into its own named function for improved readability of the parent function.

Mardak replied:
E.g., `let onIframe = iframe => {` next to the `let addTelemetry` above?

dmose commented on add-on/chrome/modules/MozLoopService.jsm:
```diff
@@ -69,6 +69,20 @@ const SHARING_SCREEN = {
+ *
+ * @type {{SHOWN: Number, NO_AGAIN: Number, NO_NEVER: Number, YES_AGAIN: Number,
+ *         YES_NEVER: Number}}
+ */
```
These all want detailed documentation, I think.  Perhaps an @enum a la http://usejsdoc.org/tags-enum.html

dmose commented:
I found the throttling code to be difficult to grok without reading through detailed comments in the bug.  I think this is actually partly do to a related issue: LoopUI is (and has been, before this patch) in danger of falling down the same rathole that MozLoopService, and, to some degree, MozLoopAPI, have fallen down: no separation of concerns.  Which is to say that we're not adhering to the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle). 

However, I think there's a pretty easy opportunity here: if we split out the throttler code into its own object, where the DNS algorithm is documented in detail and tested separately from the rest of the code, it'll be much easier to grok.  I suspect it wants to be in its own JSM, though if that's truly painful, I'm open to other possibilities.

I actually think LoopUI probably wants to be broken up even further, but we don't need to entrain that work here.  Just leading the way by making this code a clean example for future modifications would be a real win, I think.

Happy to chat about this in more detail and/or verbally as well.

Mardak commented:
Split out Throttler to a separate object in bootstrap.js with a separate mochitest. Added additional button/checkbox functionality with additional tests. All mochitests passing on a test export.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -352,6 +353,105 @@ var WindowListener = {
 
+      /**
+       * Maybe add the copy panel if it's not throttled and passes other checks.
+       * @return {Promise} Resolved when a decision and panel adding are done.
```
s/panel adding/and possibly panel-adding/ would be a bit clearer.

dmose commented on add-on/panels/js/copy.jsx:
```diff
@@ -0,0 +1,99 @@
+  function init() {
+    let { LoopAPI } = Cu.import("chrome://loop/content/modules/MozLoopAPI.jsm", {});
+    let requests = ["GetAllStrings", "GetLocale", "GetPluralRule"];
+    return Promise.all(requests.map(name => new Promise(resolve => {
```
Please do something to improve the skimmability of this, at the very least adding a comment.

dmose replied:
Splitting out a getFromLoopAPI function or something like that could be another possibility.

dmose commented on add-on/chrome/test/mochitest/browser_copypanel.js:
```diff
@@ -0,0 +1,114 @@
+
+  yield LoopUI.maybeAddCopyPanel();
+
+  Assert.equal(document.getElementById("loop-copy-notification-panel"), null, "copy panel doesn't exist already shown");
```
The failure text wants a tweak, I think.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -352,6 +353,105 @@ var WindowListener = {
+        controller.doCommand = function() {
+          // Do the normal behavior first.
+          this._doCommand.apply(this, arguments);
+
```
Let's use "controller" instead of "this" for clarity.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -926,6 +1026,7 @@ var WindowListener = {
 
     LoopUI.init();
+    window.LoopThrottler = Throttler;
     window.LoopUI = LoopUI;
```
If LoopThrottler is only exported for testing, please be explicit about that here.

dmose commented on add-on/chrome/test/mochitest/browser_copypanel.js:
```diff
@@ -0,0 +1,114 @@
+  return new Promise(resolve => {
+    iframe.addEventListener("DOMContentLoaded", () => {
+      iframe.parentNode.addEventListener("popuphidden", () => resolve(histogram));
+      setTimeout(() => onIframe(iframe));
```
Let's add a comment explaining why calling without an interval is necessary here.

dmose commented on add-on/chrome/test/mochitest/browser_copypanel.js:
```diff
@@ -0,0 +1,114 @@
+
+// Helper function for testing clicks on the copy panel.
+function testClick(domain, onIframe) {
+  let histogram = Services.telemetry.getHistogramById("LOOP_COPY_PANEL_ACTIONS");
```
Adding a comment explaining why domain is here (for test debuggability) would be great.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -968,6 +1070,75 @@ var WindowListener = {
+      let onDNS = (request, record) => {
+        // Use a specific part of the A-record IP address depending on the
+        // channel. I.e., 127.[release/beta/other].[aurora].[nightly].
+        let index = 1;
```
It sounds like we want the split to be different, with release & beta each getting their own channel, IIRC.

Mardak replied:
Yup. Fixed.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -968,6 +1070,75 @@ var WindowListener = {
+      }
+
+      // Handle responses from the DNS resolution service request.
+      let onDNS = (request, record) => {
```
Since we've got a much smaller object here, I'd propose that onDNS move to a top-level function, rather than something lexically scoped, for readability.

Mardak replied:
onDNS makes use of a couple variables in scope: `ticket` and `prefPrefix` so moving onDNS out would require some passing around those variables to a makeOnDNS function to return an onDNS callback.

dmose commented on add-on/chrome/bootstrap.js:
```diff
@@ -968,6 +1070,75 @@ var WindowListener = {
+        }
+
+        // Check if the 8-bit threshold (0-255) exceeds the ticket (0-254).
+        let threshold = record && record.getNextAddrAsString().split(".")[index];
```
While exceedingly elegant, this line has more than one layer of abstractions, and depends on knowing that IPv4 addresses when passed to split this way will create the array you want.  Please at least split this up with temporaries for readability.

dmose commented on add-on/chrome/test/mochitest/browser_copypanel.js:
```diff
@@ -0,0 +1,114 @@
+
+  Assert.equal(document.getElementById("loop-copy-notification-panel"), null, "copy panel removed");
+  Assert.strictEqual(histogram.snapshot().counts[gConstants.COPY_PANEL.NO_NEVER], 1, "triggered telemetry count for no never");
+});
```
There's nothing here that seems to test that the semantics of the _NEVER _AGAIN works.   It seems like that shouldn't be too easy, though possibly even easier to do in a unit-style test rather than by driving the UI in an integration test.  Either strategy could be added to this file, I'd say.

Mardak replied:
Added a test that tries to show the copy panel again and makes sure it doesn't.

dmose commented on add-on/chrome/test/mochitest/browser_throttler.js:
```diff
@@ -0,0 +1,167 @@
+/* Any copyright is dedicated to the Public Domain.
+   http://creativecommons.org/publicdomain/zero/1.0/ */
+
```
For bonus points, port this to xpcshell so that it runs faster.  I don't think we want to block on that if it's non-trivial, though, so at least put an XXX comment suggesting that in here.

dmose replied:
These unit tests are nicely done; good work!

Mardak replied:
Throttler is created in bootstrap.js and not a service, so it would be tricky to test via xpcshell.

dmose commented on add-on/panels/css/copy.css:
```diff
@@ -0,0 +1,49 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Nice clean CSS too!

dmose commented on add-on/panels/js/copy.jsx:
```diff
@@ -0,0 +1,99 @@
+    },
+
+    _dispatch(accept) {
+      window.dispatchEvent(new CustomEvent("CopyPanelClick", {
```
A comment describing what this event is intended to cause the chrome code to do would be nice.

dmose commented on add-on/panels/js/copy.jsx:
```diff
@@ -0,0 +1,99 @@
+
+/* global Components */
+let { utils: Cu } = Components;
+
```
If you could try setting up some simple unit tests here stubbing out the chrome things, that'd be great.  That said, I'd suggest timeboxing this so that we can still try and land tomorrow.

Mardak replied:
Stubbed as fake-components now used for desktop local tests, karma, and ui showcase.

dmose commented on add-on/panels/js/copy.jsx:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * Copy panel view.
+   */
+  let CopyView = React.createClass({
```
You should be able to use the [PureRenderMixin](https://facebook.github.io/react/docs/pure-render-mixin.html) here for a potential slight perf boost.

dmose commented:
r=dmose.  That said, if you have any code you'd like me to review before landing, I'm happy to do that.
